### ISSUE TITLE
dependency resolution: qtpy vs PySide or better: pth

### ISSUE THREAD

**Issue by Andrei-Pozolotin** [ostr00000/freecad-stubs#8]
Hello again. I hacked around the limitations described below with my own scripts and have no problem,
 but would like to see "stubs" be more cool.  :-) Therefore:

**OPTION ONE**

currently, there is a `qtpy` dependency:
```
dependencies = [
    # 'pivy', # cannot install it in a normal way, should be installed by FreeCAD anyway
    'qtpy',
]
```

and `FreeCAD` also provides its own package proxy called `PySide`
https://github.com/FreeCAD/FreeCAD/blob/main/cMake/FreeCAD_Helpers/SetupShibokenAndPyside.cmake
which is deployed on the target system at:
`/usr/lib/freecad/Ext`

and used throughout the internal code, for example:
https://github.com/FreeCAD/FreeCAD/blob/main/src/Mod/Path/Path/Tool/Gui/BitCmd.py#L28

So: would it make sense for "stubs" to drop `qtpy` and include `PySide` form the `FreeCAD` instead?

Why? so that `templates.py` could be copied as is :-) and all other `PySide` imports resolve OK
https://github.com/ostr00000/freecad-stubs/blob/main/freecad_stubs/FreeCADTemplates/templates.py

**OPTION TWO**

Alternatively, "stubs" could go even further, and instead run `FreeCAD` folder discovery, 
and generate during `pip install` automagically a `freecad.pth`  file:
```
${PROJECT_DIR}/.venv/lib/python3.11/site-packages/freecad.pth
```
with content similar to this:
```
/usr/lib/freecad/lib
/usr/lib/freecad/Ext
/usr/lib/freecad/Mod/AddonManager
```

that would also take care of `dll/so` dependency, so that tests of the plugin-under-development could run from the same `venv` against the production `FreeCAD` installation

more info on the `pth` files:
https://docs.python.org/3/library/site.html

**Comment by ostr00000** [comment]
Hi, sorry for late response.

> and `FreeCAD` also provides its own package proxy called `PySide`

I did not know that. Thanks for info.

## OPTION ONE

- `-` I see one big advantage `qypt` to `PySide` - type checkers support. There is support in [`qtpy`](https://github.com/spyder-ide/qtpy#type-checker-integration) for both most popular type checkers (`mypy` and `pyright`).
Maybe it would be possible to add typing support (after all, it only redirects imports). But this should be done directly in FreeCAD repo. Probably `py.types` marker is [enough](https://peps.python.org/pep-0561/#packaging-type-information).
- `-` Another advantage for `qypt` is wider `qt` version compatibility. Sometimes `qt` change destination module for some class or change method name. Usually `qtpy` has both method available (ex. redirect import, adds method alias), so code has longer support.
I see some code that try to do this in `PySide.QtGui`, but would not be easier to use already existing library for this purpose? (question to FreeCAD developers).

> So: would it make sense for "stubs" to drop `qtpy` and include `PySide` form the `FreeCAD` instead?

- `+` Maybe, I think that FreeCAD having its own compatibility module is a very good argument.
- `0` But on the other hand, it is quite easy to change to `PySide` version (just replace `qtpy` to `PySide` in imports, right? or did you have to do anything more?).
- `0` Also, installing `qtpy` should be very easy (comparing to path configuration for FreeCAD `Ext`).

>  and all other `PySide` imports resolve OK

Yes, but only when FreeCAD python environment is configured (When I generate stubs, I do not have configured environment, so I will not be able to verify if imports are corrects).

## OPTION TWO

> run `FreeCAD` folder discovery and generate during `pip install` automagically a `freecad.pth`

I do not see how it is connected to the stub package. If someone do not use stubs, then the same problem occurs.


### Alternative solution

Surly, we can create another package (ex. `freecad-pth`). Unfortunately, I can prepare FreeCAD `.pth` only for Linux based OS (operating system) (I do not have other OSs).
I am afraid that there is no single general discovery rule. Different OS may have different installation path (in theory even on Linux, no idea how it looks in practice). 
Another implementation details: it may be impossible to run post install script for wheel - as far as I know, it is by design. It is a minor problem, because it may be done in 2-phases: 
1. installing library (ex. `pip install freecad-pth`),
2. and running proper program (ex. `freecad-pth`).

So it clearly visible that this is a package maintainer responsibility to manage these paths. (BTW, I see that deb packages have finally a [new maintainer](https://forum.freecad.org/viewtopic.php?p=723372#p723372)). 
On the other hand, there probably is a good reason why this is not done today (maybe to not polluting `PYTHONPATH` outside FreeCAD?)

### Adding FreeCAD path to `PYTHONPATH`

Here is a `bash` script I use to make FreeCAD visible to `python` executable - may be useful to somebody who look at this thread.
```bash
function join() {
  local IFS=$1
  shift
  echo "$*"
}
PYTHONPATH="/usr/lib/freecad-python3/lib/:/usr/share/freecad/Ext/:$(join ':' /usr/share/freecad/Mod/*):$PYTHONPATH" python3
```

### Generating `freecad.pth`
And here is a script for generating `pth` file:
```bash
python -c 'import sys; [print(a) for a in sys.argv[1:]]' /usr/lib/freecad-python3/lib/ /usr/share/freecad/Ext/ /usr/share/freecad/Mod/* > freecad.pth
```
Or directly to current site/dist package folder (need to have correct permissions):
```bash
python -c 'import itertools, pathlib, sys, site; (p := pathlib.Path(site.getsitepackages()[0] + "/freecad.pth")).write_text("\n".join(itertools.chain(("/usr/lib/freecad-python3/lib/", "/usr/share/freecad/Ext/"), (str(p) for p in pathlib.Path("/usr/share/freecad/Mod").glob("*"))))); print(f"\nFreeCAD python paths saved in: {p}")'
```

## Next steps for this issue
- Continue discussion: 
@Andrei-Pozolotin please tell me if these arguments are convincing or not.

- Documenting `PYTHONPATH`: 
I found recently [FreeCAD Handbook](https://github.com/FreeCAD/DevelopersHandbook). Although it has mostly C++ tips, there are also some Python sections. It may be worth to add there above info about `PYTOHNPATH`.

**Comment by Andrei-Pozolotin** [comment]
Thank you for taking your time to consider this. You bring up good valid points, now I am confused and have no position on any of this  , perhaps need more time to "meditate" :heart_eyes:.

Meanwhile, one more simple idea overlooked before: use the usual "multi-import hack" in `templates.py` to support copy-as-is approach:

```python
QT_has_missing_import = True
if QT_has_missing_import:
    try:
        from qtpy.QtCore import QObject
        from qtpy.QtGui import QIcon
        from qtpy.QtWidgets import QDialogButtonBox, QMenu
        QT_has_missing_import = False
    except:
        QT_has_missing_import = True
if QT_has_missing_import:
    try:
        from PySide.QtCore import QObject
        from PySide.QtGui import QIcon
        from PySide.QtWidgets import QDialogButtonBox, QMenu
        QT_has_missing_import = False
    except:
        QT_has_missing_import = True
assert not QT_has_missing_import, "Missing Python-QT Bridge"
del QT_has_missing_import
```

**Comment by ostr00000** [comment]
I like solution with `try except`. It should make code independent of proxy library.
I fixed it in https://github.com/ostr00000/freecad-stubs/commit/09e2608b859428ac024df6630603d726adab397e#diff-82c5b3347e6faf6141824b74aff7eb9081cdb124038d1b7f4c96b4556a550bfeR25-R45
I used solutions from https://github.com/spyder-ide/qtpy/issues/447 to use `if` block for `pyright` conditionally.

Also, `pre-commit` detected that all this import section does not have to be imported at all, if we use `from __future__ import annotations`.
